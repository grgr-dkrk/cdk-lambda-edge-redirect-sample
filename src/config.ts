import * as dotenv from 'dotenv'

dotenv.config()

/**
 * account & region
 */
export const AWS_ACCOUNT = process.env.AWS_ACCOUNT as string
export const AWS_REGION = process.env.AWS_REGION ?? 'us-east-1'
export const env = {
  account: AWS_ACCOUNT,
  region: AWS_REGION,
}

/**
 * Environment
 */
export const ENVIRONMENT_STAGING = 'staging' as const
export const ENVIRONMENT_PRODUCTION = 'production' as const
export type Environment =
  | typeof ENVIRONMENT_STAGING
  | typeof ENVIRONMENT_PRODUCTION

/**
 * Domain
 */
export const DOMAIN_STAGING = process.env.DOMAIN_STAGING as string
export const DOMAIN_PRODUCTION = process.env.DOMAIN_PRODUCTION as string
export type DomainRecord = Record<
  typeof DOMAIN_STAGING | typeof DOMAIN_PRODUCTION,
  string
>
export const domain: DomainRecord = {
  [ENVIRONMENT_STAGING]: DOMAIN_STAGING,
  [ENVIRONMENT_PRODUCTION]: DOMAIN_PRODUCTION,
}

/**
 * HostedZone
 */
export const HOSTED_ZONE_STAGING = process.env.HOSTED_ZONE_STAGING as string
export const HOSTED_ZONE_PRODUCTION = process.env
  .HOSTED_ZONE_PRODUCTION as string
export const hostedZone: Record<Environment, string> = {
  [ENVIRONMENT_STAGING]: HOSTED_ZONE_STAGING,
  [ENVIRONMENT_PRODUCTION]: HOSTED_ZONE_PRODUCTION,
}

/**
 * Policy
 */
export const USER_ARN_STAGING = process.env.USER_ARN_STAGING as string
export const USER_ARN_PRODUCTION = process.env.USER_ARN_PRODUCTION as string
export const principal: Record<Environment, string> = {
  [ENVIRONMENT_STAGING]: USER_ARN_STAGING,
  [ENVIRONMENT_PRODUCTION]: USER_ARN_PRODUCTION,
}

/**
 * Assets
 */
export const ASSETS_WEBSITE = './website' as const
