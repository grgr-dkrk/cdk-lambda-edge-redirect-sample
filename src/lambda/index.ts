import { CloudFrontRequestEvent, CloudFrontRequestResult } from 'aws-lambda'
import { normalizeUri } from './normalizeUri'
import { checkRedirectResponse } from './redirect'

export const lambdaFunctions = async (
  event: CloudFrontRequestEvent,
): Promise<CloudFrontRequestResult> => {
  const request = event.Records[0].cf.request
  const { uri } = request

  // redirect
  const redirectResponse = checkRedirectResponse(uri)
  if (redirectResponse) {
    return redirectResponse
  }

  // rewrite
  const normalizedUri = normalizeUri(uri)
  return { ...request, uri: normalizedUri }
}
