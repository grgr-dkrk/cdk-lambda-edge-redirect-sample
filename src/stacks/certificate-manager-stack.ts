import { StackProps, Stack, Construct } from '@aws-cdk/core'
import { DnsValidatedCertificate } from '@aws-cdk/aws-certificatemanager'
import { IHostedZone } from '@aws-cdk/aws-route53'

type Props = StackProps & {
  hostedZoneName: string
  zone: IHostedZone
}

export class CertificateManagerStack extends Stack {
  public readonly certificate: DnsValidatedCertificate
  constructor(scope: Construct, id: string, props: Props) {
    super(scope, id, props)
    /**
     * environment, domain
     */
    const { zone, hostedZoneName } = props

    /**
     * Certificate
     */
    this.certificate = new DnsValidatedCertificate(this, 'Certificate', {
      domainName: hostedZoneName,
      subjectAlternativeNames: [`*.${zone.zoneName}`],
      region: 'us-east-1',
      hostedZone: zone,
    })
  }
}
