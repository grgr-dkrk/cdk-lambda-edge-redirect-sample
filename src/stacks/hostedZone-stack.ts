import { StackProps, Stack, Construct } from '@aws-cdk/core'
import { HostedZone, IHostedZone } from '@aws-cdk/aws-route53'

type Props = StackProps & {
  hostedZoneName: string
}

export class HostedZoneStack extends Stack {
  public readonly zone: IHostedZone
  constructor(scope: Construct, id: string, props: Props) {
    super(scope, id, props)
    /**
     * environment, domain,
     */
    const { hostedZoneName } = props

    /**
     * HostedZone
     */
    this.zone = HostedZone.fromLookup(this, 'HostedZone', {
      domainName: hostedZoneName,
    })
  }
}
