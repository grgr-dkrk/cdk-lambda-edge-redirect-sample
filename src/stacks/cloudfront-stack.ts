import { Bucket } from '@aws-cdk/aws-s3'
import { RemovalPolicy, StackProps, Stack, Construct } from '@aws-cdk/core'
import { Environment } from '../config'
import { DnsValidatedCertificate } from '@aws-cdk/aws-certificatemanager'
import {
  IHostedZone,
  ARecord,
  AaaaRecord,
  RecordTarget,
  ARecordProps,
} from '@aws-cdk/aws-route53'
import {
  OriginAccessIdentity,
  CloudFrontWebDistribution,
  ViewerCertificate,
  SecurityPolicyProtocol,
  SSLMethod,
  LambdaEdgeEventType,
} from '@aws-cdk/aws-cloudfront'
import { CloudFrontTarget } from '@aws-cdk/aws-route53-targets'
import {
  AssetCode,
  Function as LambdaFunction,
  Runtime,
} from '@aws-cdk/aws-lambda'
import { resolve } from 'path'

type Props = StackProps & {
  environment: Environment
  domainName: string
  certificate: DnsValidatedCertificate
  zone: IHostedZone
}

export class CloudFrontStack extends Stack {
  public readonly bucket: Bucket
  public readonly distribution: CloudFrontWebDistribution

  constructor(scope: Construct, id: string, props: Props) {
    super(scope, id, props)
    /**
     * environment, domain
     */
    const { certificate, domainName, environment, zone } = props

    /**
     * OAI
     */
    const originAccessIdentity = new OriginAccessIdentity(this, 'Oai')

    /**
     * Lambda
     */
    const lambdaCode = new AssetCode(`${resolve(__dirname)}/../lambda/dist`)
    const lambdaFunction = new LambdaFunction(this, 'redirect', {
      handler: 'index.lambdaFunctions',
      code: lambdaCode,
      runtime: Runtime.NODEJS_12_X,
    }).currentVersion

    /**
     * Bucket
     */
    this.bucket = new Bucket(this, `Bucket${id}`, {
      bucketName: `lambda-edge-redirect-${environment}`,
      removalPolicy: RemovalPolicy.DESTROY,
      websiteIndexDocument: 'index.html',
    })

    /**
     * Set GrantRead to Distribution
     */
    this.bucket.grantRead(originAccessIdentity)

    /**
     * Distribution
     */
    this.distribution = new CloudFrontWebDistribution(this, `Distribution`, {
      originConfigs: [
        {
          s3OriginSource: {
            s3BucketSource: this.bucket,
            originAccessIdentity,
          },
          behaviors: [
            {
              compress: true,
              isDefaultBehavior: true,
              lambdaFunctionAssociations: [
                {
                  eventType: LambdaEdgeEventType.ORIGIN_REQUEST,
                  lambdaFunction,
                },
              ],
            },
          ],
        },
      ],
      viewerCertificate: ViewerCertificate.fromAcmCertificate(certificate, {
        aliases: [domainName],
        securityPolicy: SecurityPolicyProtocol.TLS_V1,
        sslMethod: SSLMethod.SNI,
      }),
    })

    /**
     * Set ARecord
     */
    const route53RecordProps: ARecordProps = {
      zone,
      recordName: domainName,
      target: RecordTarget.fromAlias(new CloudFrontTarget(this.distribution)),
    }
    new ARecord(this, 'ARecord', route53RecordProps)
    new AaaaRecord(this, 'AaaaRecord', route53RecordProps)
  }
}
