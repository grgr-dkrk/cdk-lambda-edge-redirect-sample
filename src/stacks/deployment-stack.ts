import { StackProps, Stack, Construct } from '@aws-cdk/core'
import { ISource, BucketDeployment } from '@aws-cdk/aws-s3-deployment'
import { CloudFrontStack } from './cloudfront-stack'

type Props = StackProps & {
  cloudFront: CloudFrontStack
  source: ISource
}

export class DeploymentStack extends Stack {
  constructor(scope: Construct, id: string, props: Props) {
    super(scope, id, props)
    const { cloudFront, source } = props
    const { bucket, distribution } = cloudFront

    /**
     * Bucket Deployment
     */
    new BucketDeployment(this, `BucketDeployment${id}`, {
      destinationBucket: bucket,
      sources: [source],
      distribution,
    })
  }
}
