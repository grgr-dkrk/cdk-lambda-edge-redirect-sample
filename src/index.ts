#!/usr/bin/env node
import 'source-map-support/register'
import { App } from '@aws-cdk/core'
import { DeploymentStack } from './stacks/deployment-stack'
import { Environment, env, domain, hostedZone } from './config'
import { HostedZoneStack } from './stacks/hostedZone-stack'
import { CertificateManagerStack } from './stacks/certificate-manager-stack'
import { CloudFrontStack } from './stacks/cloudfront-stack'
import { resolve } from 'path'
import { Source } from '@aws-cdk/aws-s3-deployment'

const app = new App()

const environment = app.node.tryGetContext('environment') as
  | Environment
  | undefined

if (!env.account) throw new Error('not found: account')
if (!env.region) throw new Error('not found: region')
if (!environment) throw new Error('not found: environment')

const domainName = domain[environment]
const hostedZoneName = hostedZone[environment]

/**
 * HostedZoneStack
 */
const { zone } = new HostedZoneStack(app, `HostedZoneStack-${environment}`, {
  hostedZoneName,
  env,
})

/**
 * CertificateManagerStack
 */
const { certificate } = new CertificateManagerStack(
  app,
  `CertificateManagerStack-${environment}`,
  {
    hostedZoneName,
    zone,
    env,
  },
)

/**
 * CloudFrontStackStack
 */
const cloudFront = new CloudFrontStack(app, `CloudFrontStack-${environment}`, {
  environment,
  domainName,
  certificate,
  zone,
  env,
})

/**
 * DeploymentStack
 */
new DeploymentStack(app, `DeploymentStack-${environment}`, {
  cloudFront,
  source: Source.asset(`${resolve(__dirname)}/website/`),
  env,
})
