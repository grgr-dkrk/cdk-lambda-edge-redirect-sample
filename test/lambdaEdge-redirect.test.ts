import { SynthUtils } from '@aws-cdk/assert'
import * as cdk from '@aws-cdk/core'
import * as CdkAlbRedirect from '../src/stacks/deployment-stack'
import { ENVIRONMENT_STAGING } from '../src/config'
import { HostedZoneStack } from '../src/stacks/hostedZone-stack'
import { CertificateManagerStack } from '../src/stacks/certificate-manager-stack'
import { env } from './__mocks__/env'
import { CloudFrontStack } from '../src/stacks/cloudfront-stack'
import { Source } from '@aws-cdk/aws-s3-deployment'
import { resolve } from 'path'

it('snapshot: CdkDeploymentStack', () => {
  const app = new cdk.App()
  const environment = ENVIRONMENT_STAGING
  const domainName = 'dkrk-blog.net'
  const hostedZoneName = 'dkrk-blog.net'

  const { zone } = new HostedZoneStack(app, 'testDns', {
    hostedZoneName,
    env,
  })

  const { certificate } = new CertificateManagerStack(app, 'testCert', {
    hostedZoneName,
    zone,
    env,
  })

  const cloudFront = new CloudFrontStack(app, 'testCf', {
    environment,
    domainName,
    certificate,
    zone,
    env,
  })

  const stack = new CdkAlbRedirect.DeploymentStack(app, 'MyTestStack', {
    cloudFront,
    source: Source.asset(`${resolve(__dirname)}/__mocks__/sources`),
    env,
  })

  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot()
})
