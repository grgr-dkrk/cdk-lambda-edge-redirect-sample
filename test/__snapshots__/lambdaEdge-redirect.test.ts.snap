// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot: CdkDeploymentStack 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters4184245adc1f2ed71e1f0ae5719f8fd7f34324b750f1bf06b2fb5cf1f4014f50ArtifactHash846130E4": Object {
      "Description": "Artifact hash for asset \\"4184245adc1f2ed71e1f0ae5719f8fd7f34324b750f1bf06b2fb5cf1f4014f50\\"",
      "Type": "String",
    },
    "AssetParameters4184245adc1f2ed71e1f0ae5719f8fd7f34324b750f1bf06b2fb5cf1f4014f50S3BucketB6159468": Object {
      "Description": "S3 bucket for asset \\"4184245adc1f2ed71e1f0ae5719f8fd7f34324b750f1bf06b2fb5cf1f4014f50\\"",
      "Type": "String",
    },
    "AssetParameters4184245adc1f2ed71e1f0ae5719f8fd7f34324b750f1bf06b2fb5cf1f4014f50S3VersionKey2060CAC0": Object {
      "Description": "S3 key for asset version \\"4184245adc1f2ed71e1f0ae5719f8fd7f34324b750f1bf06b2fb5cf1f4014f50\\"",
      "Type": "String",
    },
    "AssetParameters4c45d5b7d85dc011ab7bcb6e67c053fb49e9dd4b28a0461b8c443bd12a02c42fArtifactHash3E423712": Object {
      "Description": "Artifact hash for asset \\"4c45d5b7d85dc011ab7bcb6e67c053fb49e9dd4b28a0461b8c443bd12a02c42f\\"",
      "Type": "String",
    },
    "AssetParameters4c45d5b7d85dc011ab7bcb6e67c053fb49e9dd4b28a0461b8c443bd12a02c42fS3Bucket1F278211": Object {
      "Description": "S3 bucket for asset \\"4c45d5b7d85dc011ab7bcb6e67c053fb49e9dd4b28a0461b8c443bd12a02c42f\\"",
      "Type": "String",
    },
    "AssetParameters4c45d5b7d85dc011ab7bcb6e67c053fb49e9dd4b28a0461b8c443bd12a02c42fS3VersionKey691264E3": Object {
      "Description": "S3 key for asset version \\"4c45d5b7d85dc011ab7bcb6e67c053fb49e9dd4b28a0461b8c443bd12a02c42f\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "BucketDeploymentMyTestStackCustomResource9A69C40D": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "DestinationBucketName": Object {
          "Fn::ImportValue": "testCf:ExportsOutputRefBuckettestCfFA6CFF1D52835980",
        },
        "DistributionId": Object {
          "Fn::ImportValue": "testCf:ExportsOutputRefDistributionCFDistribution882A731330275A1E",
        },
        "Prune": true,
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536",
            "Arn",
          ],
        },
        "SourceBucketNames": Array [
          Object {
            "Ref": "AssetParameters4c45d5b7d85dc011ab7bcb6e67c053fb49e9dd4b28a0461b8c443bd12a02c42fS3Bucket1F278211",
          },
        ],
        "SourceObjectKeys": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters4c45d5b7d85dc011ab7bcb6e67c053fb49e9dd4b28a0461b8c443bd12a02c42fS3VersionKey691264E3",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters4c45d5b7d85dc011ab7bcb6e67c053fb49e9dd4b28a0461b8c443bd12a02c42fS3VersionKey691264E3",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        ],
      },
      "Type": "Custom::CDKBucketDeployment",
      "UpdateReplacePolicy": "Delete",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756C81C01536": Object {
      "DependsOn": Array [
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters4184245adc1f2ed71e1f0ae5719f8fd7f34324b750f1bf06b2fb5cf1f4014f50S3BucketB6159468",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters4184245adc1f2ed71e1f0ae5719f8fd7f34324b750f1bf06b2fb5cf1f4014f50S3VersionKey2060CAC0",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters4184245adc1f2ed71e1f0ae5719f8fd7f34324b750f1bf06b2fb5cf1f4014f50S3VersionKey2060CAC0",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
            "Arn",
          ],
        },
        "Runtime": "python3.6",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "AssetParameters4c45d5b7d85dc011ab7bcb6e67c053fb49e9dd4b28a0461b8c443bd12a02c42fS3Bucket1F278211",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "AssetParameters4c45d5b7d85dc011ab7bcb6e67c053fb49e9dd4b28a0461b8c443bd12a02c42fS3Bucket1F278211",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::ImportValue": "testCf:ExportsOutputFnGetAttBuckettestCfFA6CFF1DArn01E2EF58",
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      Object {
                        "Fn::ImportValue": "testCf:ExportsOutputFnGetAttBuckettestCfFA6CFF1DArn01E2EF58",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": Array [
                "cloudfront:GetInvalidation",
                "cloudfront:CreateInvalidation",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRoleDefaultPolicy88902FDF",
        "Roles": Array [
          Object {
            "Ref": "CustomCDKBucketDeployment8693BB64968944B69AAFB0CC9EB8756CServiceRole89A01265",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
}
`;
